File: backend/app/schemas/smart_matching.py
"""
Pydantic schemas for Smart Matching Engine API
"""

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from datetime import datetime
from enum import Enum

from .base import BaseSchema


class MatchConfidenceLevel(str, Enum):
    EXCELLENT = "EXCELLENT"
    VERY_GOOD = "VERY_GOOD"
    GOOD = "GOOD"
    FAIR = "FAIR"
    POOR = "POOR"


class MatchTypeEnum(str, Enum):
    ORDER_TO_PRODUCTION_QUOTE = "order_to_production_quote"
    PRODUCTION_QUOTE_TO_ORDER = "production_quote_to_order"


class FeedbackType(str, Enum):
    HELPFUL = "helpful"
    NOT_HELPFUL = "not_helpful"
    CONTACTED = "contacted"
    CONVERTED = "converted"
    IRRELEVANT = "irrelevant"


class SmartMatchingRequest(BaseModel):
    """Request model for smart matching recommendations"""
    order_id: int = Field(..., description="Order ID to generate recommendations for")
    max_recommendations: int = Field(15, ge=1, le=50, description="Maximum number of recommendations")
    include_ai_insights: bool = Field(True, description="Include AI-powered insights")
    enable_ml_predictions: bool = Field(True, description="Enable machine learning predictions")
    custom_weights: Optional[Dict[str, float]] = Field(None, description="Custom scoring weights")
    filters: Optional[Dict[str, Any]] = Field(None, description="Additional filters")


class MatchScoreBreakdown(BaseModel):
    """Detailed match score breakdown"""
    total_score: float = Field(..., ge=0, le=1, description="Overall match score")
    capability_score: float = Field(..., ge=0, le=1, description="Capability matching score")
    performance_score: float = Field(..., ge=0, le=1, description="Performance history score")
    geographic_score: float = Field(..., ge=0, le=1, description="Geographic proximity score")
    quality_score: float = Field(..., ge=0, le=1, description="Quality metrics score")
    cost_efficiency_score: float = Field(..., ge=0, le=1, description="Cost efficiency score")
    availability_score: float = Field(..., ge=0, le=1, description="Availability score")
    confidence_level: float = Field(..., ge=0, le=1, description="Confidence in the match")
    recommendation_strength: str = Field(..., description="STRONG, MODERATE, or WEAK")
    match_reasons: List[str] = Field(default_factory=list, description="Reasons for the match")
    risk_factors: List[str] = Field(default_factory=list, description="Potential risk factors")


class SmartRecommendationData(BaseModel):
    """Individual smart recommendation data"""
    manufacturer_id: int = Field(..., description="Manufacturer ID")
    manufacturer_name: str = Field(..., description="Manufacturer business name")
    match_score: MatchScoreBreakdown = Field(..., description="Detailed match score")
    predicted_success_rate: float = Field(..., ge=0, le=1, description="AI-predicted success rate")
    estimated_delivery_time: int = Field(..., description="Estimated delivery time in days")
    estimated_cost_range: Dict[str, float] = Field(..., description="Estimated cost range")
    risk_assessment: Dict[str, Any] = Field(..., description="Risk assessment details")
    competitive_advantages: List[str] = Field(default_factory=list, description="Key advantages")
    potential_concerns: List[str] = Field(default_factory=list, description="Potential concerns")
    similar_past_projects: List[Dict[str, Any]] = Field(default_factory=list, description="Similar projects")
    ai_insights: Dict[str, Any] = Field(default_factory=dict, description="AI-generated insights")


class SmartMatchingResponse(BaseModel):
    """Response model for smart matching recommendations"""
    success: bool = Field(..., description="Request success status")
    message: str = Field(..., description="Response message")
    order_id: int = Field(..., description="Order ID")
    recommendations_count: int = Field(..., description="Number of recommendations generated")
    processing_time_seconds: float = Field(..., description="Processing time in seconds")
    algorithm_version: str = Field(..., description="Algorithm version used")
    recommendations: List[SmartRecommendationData] = Field(..., description="Smart recommendations")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="Additional metadata")


class RecommendationResponse(BaseModel):
    """Simplified recommendation response"""
    manufacturer_id: int = Field(..., description="Manufacturer ID")
    manufacturer_name: str = Field(..., description="Manufacturer name")
    total_score: float = Field(..., ge=0, le=1, description="Total match score")
    confidence_level: float = Field(..., ge=0, le=1, description="Confidence level")
    recommendation_strength: str = Field(..., description="Recommendation strength")
    predicted_success_rate: float = Field(..., ge=0, le=1, description="Predicted success rate")
    estimated_delivery_days: int = Field(..., description="Estimated delivery days")
    estimated_cost_range: Dict[str, float] = Field(..., description="Cost range estimate")
    key_advantages: List[str] = Field(default_factory=list, description="Key advantages")
    potential_risks: List[str] = Field(default_factory=list, description="Potential risks")
    match_summary: str = Field("", description="AI-generated match summary")


class MatchAnalysisRequest(BaseModel):
    """Request for detailed match analysis"""
    order_id: int = Field(..., description="Order ID")
    manufacturer_id: int = Field(..., description="Manufacturer ID")
    analysis_depth: str = Field("detailed", description="Analysis depth: basic, detailed, comprehensive")


class MatchAnalysisResponse(BaseModel):
    """Response for detailed match analysis"""
    order_id: int = Field(..., description="Order ID")
    manufacturer_id: int = Field(..., description="Manufacturer ID")
    analysis: Dict[str, Any] = Field(..., description="Detailed analysis results")
    generated_at: datetime = Field(..., description="Analysis generation timestamp")


class AIInsightsResponse(BaseModel):
    """AI insights response"""
    order_id: int = Field(..., description="Order ID")
    insights: Dict[str, Any] = Field(..., description="AI-generated insights")
    confidence_score: float = Field(..., ge=0, le=1, description="Confidence in insights")
    generated_at: datetime = Field(..., description="Generation timestamp")


class ManufacturerCluster(BaseModel):
    """Manufacturer cluster information"""
    cluster_id: int = Field(..., description="Cluster ID")
    cluster_name: str = Field(..., description="Cluster name")
    description: str = Field(..., description="Cluster description")
    manufacturer_count: int = Field(..., description="Number of manufacturers in cluster")
    avg_rating: float = Field(..., description="Average rating in cluster")
    key_capabilities: List[str] = Field(..., description="Key capabilities")
    typical_industries: List[str] = Field(..., description="Typical industries served")


class ManufacturerClusterResponse(BaseModel):
    """Manufacturer clustering response"""
    clusters: List[ManufacturerCluster] = Field(..., description="Manufacturer clusters")
    total_manufacturers: int = Field(..., description="Total manufacturers analyzed")
    analysis_date: datetime = Field(..., description="Analysis date")
    clustering_algorithm: str = Field(..., description="Clustering algorithm used")


class MarketIntelligenceResponse(BaseModel):
    """Market intelligence response"""
    intelligence: Dict[str, Any] = Field(..., description="Market intelligence data")
    timeframe: str = Field(..., description="Analysis timeframe")
    region: str = Field(..., description="Geographic region")
    industry: str = Field(..., description="Industry focus")
    generated_at: datetime = Field(..., description="Generation timestamp")
    data_freshness_hours: int = Field(..., description="Data freshness in hours")


class MatchingFeedback(BaseModel):
    """Matching feedback model"""
    order_id: int = Field(..., description="Order ID")
    manufacturer_id: int = Field(..., description="Manufacturer ID")
    feedback_type: str = Field(..., description="Type of feedback")
    rating: int = Field(..., ge=1, le=5, description="Rating from 1-5")
    comments: Optional[str] = Field(None, description="Additional comments")


class MatchingAnalytics(BaseModel):
    """Matching analytics data"""
    total_matches_generated: int = Field(..., description="Total matches generated")
    avg_processing_time: float = Field(..., description="Average processing time")
    success_rate: float = Field(..., description="Matching success rate")
    user_satisfaction: float = Field(..., description="User satisfaction score")
    algorithm_performance: Dict[str, float] = Field(..., description="Algorithm performance metrics")


class SmartMatchingConfig(BaseModel):
    """Smart matching configuration"""
    weights: Dict[str, float] = Field(..., description="Scoring weights")
    thresholds: Dict[str, float] = Field(..., description="Matching thresholds")
    ml_enabled: bool = Field(True, description="Machine learning enabled")
    clustering_enabled: bool = Field(True, description="Clustering enabled")
    max_recommendations: int = Field(15, description="Maximum recommendations")


class BatchMatchingRequest(BaseModel):
    """Batch matching request for multiple orders"""
    order_ids: List[int] = Field(..., description="List of order IDs")
    max_recommendations_per_order: int = Field(10, description="Max recommendations per order")
    include_ai_insights: bool = Field(False, description="Include AI insights")
    priority: str = Field("normal", description="Processing priority: low, normal, high")


class BatchMatchingResponse(BaseModel):
    """Batch matching response"""
    success: bool = Field(..., description="Batch processing success")
    processed_orders: int = Field(..., description="Number of orders processed")
    total_recommendations: int = Field(..., description="Total recommendations generated")
    processing_time_seconds: float = Field(..., description="Total processing time")
    results: List[SmartMatchingResponse] = Field(..., description="Individual results")
    errors: List[Dict[str, Any]] = Field(default_factory=list, description="Processing errors")


class RealtimeMatchingUpdate(BaseModel):
    """Real-time matching update"""
    order_id: int = Field(..., description="Order ID")
    update_type: str = Field(..., description="Update type")
    new_recommendations: List[SmartRecommendationData] = Field(default_factory=list)
    removed_recommendations: List[int] = Field(default_factory=list)
    score_updates: Dict[int, float] = Field(default_factory=dict)
    timestamp: datetime = Field(..., description="Update timestamp")


class MatchingPerformanceMetrics(BaseModel):
    """Matching performance metrics"""
    algorithm_version: str = Field(..., description="Algorithm version")
    avg_processing_time_ms: float = Field(..., description="Average processing time in milliseconds")
    accuracy_score: float = Field(..., description="Matching accuracy score")
    precision: float = Field(..., description="Precision metric")
    recall: float = Field(..., description="Recall metric")
    f1_score: float = Field(..., description="F1 score")
    user_satisfaction_rating: float = Field(..., description="User satisfaction rating")
    conversion_rate: float = Field(..., description="Recommendation to quote conversion rate")


class SmartMatchingHealthCheck(BaseModel):
    """Smart matching system health check"""
    status: str = Field(..., description="System status: healthy, degraded, down")
    ml_models_loaded: bool = Field(..., description="ML models loaded status")
    database_connection: bool = Field(..., description="Database connection status")
    cache_status: str = Field(..., description="Cache system status")
    last_model_update: datetime = Field(..., description="Last ML model update")
    active_recommendations: int = Field(..., description="Active recommendations count")
    system_load: float = Field(..., description="Current system load")


class ExplainabilityReport(BaseModel):
    """AI explainability report for recommendations"""
    order_id: int = Field(..., description="Order ID")
    manufacturer_id: int = Field(..., description="Manufacturer ID")
    decision_factors: List[Dict[str, Any]] = Field(..., description="Key decision factors")
    feature_importance: Dict[str, float] = Field(..., description="Feature importance scores")
    similar_cases: List[Dict[str, Any]] = Field(..., description="Similar historical cases")
    confidence_intervals: Dict[str, List[float]] = Field(..., description="Confidence intervals")
    what_if_scenarios: List[Dict[str, Any]] = Field(..., description="What-if scenario analysis")
    generated_at: datetime = Field(..., description="Report generation timestamp")


# Core Match Score Schema
class MatchScoreResponse(BaseModel):
    total_score: float = Field(..., ge=0.0, le=1.0, description="Overall match score")
    category_match: float = Field(..., ge=0.0, le=1.0, description="Manufacturing category compatibility")
    price_compatibility: float = Field(..., ge=0.0, le=1.0, description="Price range compatibility")
    timeline_compatibility: float = Field(..., ge=0.0, le=1.0, description="Delivery timeline compatibility")
    geographic_proximity: float = Field(..., ge=0.0, le=1.0, description="Geographic proximity score")
    capacity_availability: float = Field(..., ge=0.0, le=1.0, description="Production capacity availability")
    manufacturer_rating: float = Field(..., ge=0.0, le=1.0, description="Manufacturer quality rating")
    urgency_alignment: float = Field(..., ge=0.0, le=1.0, description="Urgency level alignment")
    specification_match: float = Field(..., ge=0.0, le=1.0, description="Technical specification match")
    confidence_level: MatchConfidenceLevel = Field(..., description="Overall confidence in match")
    match_reasons: List[str] = Field(default_factory=list, description="Reasons for the match")
    potential_issues: List[str] = Field(default_factory=list, description="Potential concerns or issues")


# Manufacturer Info Schema
class MatchManufacturerInfo(BaseModel):
    id: int
    name: str
    location: Optional[str] = None
    rating: Optional[float] = Field(None, ge=0.0, le=5.0)
    verified: bool = False
    completed_orders: int = 0


# Main Smart Match Response
class SmartMatchResponse(BaseModel):
    match_id: str = Field(..., description="Unique match identifier")
    match_type: MatchTypeEnum = Field(..., description="Type of match")
    order_id: Optional[int] = Field(None, description="Order ID if applicable")
    production_quote_id: Optional[int] = Field(None, description="Production quote ID if applicable")
    score: MatchScoreResponse = Field(..., description="Detailed match scoring")
    estimated_price: Optional[float] = Field(None, description="Estimated price for the match")
    estimated_delivery_days: Optional[int] = Field(None, description="Estimated delivery time in days")
    manufacturer_info: MatchManufacturerInfo = Field(..., description="Manufacturer information")
    created_at: datetime = Field(..., description="When the match was generated")
    expires_at: Optional[datetime] = Field(None, description="When the match expires")
    real_time_availability: Optional[str] = Field(None, description="Real-time availability status")

    @classmethod
    def from_smart_match(cls, smart_match):
        """Convert SmartMatch service object to response schema."""
        return cls(
            match_id=smart_match.match_id,
            match_type=smart_match.match_type.value,
            order_id=smart_match.order_id,
            production_quote_id=smart_match.production_quote_id,
            score=MatchScoreResponse(
                total_score=smart_match.score.total_score,
                category_match=smart_match.score.category_match,
                price_compatibility=smart_match.score.price_compatibility,
                timeline_compatibility=smart_match.score.timeline_compatibility,
                geographic_proximity=smart_match.score.geographic_proximity,
                capacity_availability=smart_match.score.capacity_availability,
                manufacturer_rating=smart_match.score.manufacturer_rating,
                urgency_alignment=smart_match.score.urgency_alignment,
                specification_match=smart_match.score.specification_match,
                confidence_level=smart_match.score.confidence_level,
                match_reasons=smart_match.score.match_reasons,
                potential_issues=smart_match.score.potential_issues
            ),
            estimated_price=smart_match.estimated_price,
            estimated_delivery_days=smart_match.estimated_delivery_days,
            manufacturer_info=MatchManufacturerInfo(**smart_match.manufacturer_info),
            created_at=smart_match.created_at,
            expires_at=smart_match.expires_at,
            real_time_availability=getattr(smart_match, 'real_time_availability', None)
        )


# Filter and Search Schemas
class MatchFilters(BaseModel):
    min_score: Optional[float] = Field(0.6, ge=0.0, le=1.0, description="Minimum match score")
    max_price: Optional[float] = Field(None, gt=0, description="Maximum acceptable price")
    max_delivery_days: Optional[int] = Field(None, gt=0, description="Maximum delivery time in days")
    required_categories: Optional[List[str]] = Field(None, description="Required manufacturing categories")
    preferred_locations: Optional[List[str]] = Field(None, description="Preferred geographic locations")
    min_manufacturer_rating: Optional[float] = Field(None, ge=0.0, le=5.0, description="Minimum manufacturer rating")
    verified_only: bool = Field(False, description="Only include verified manufacturers")
    confidence_levels: Optional[List[MatchConfidenceLevel]] = Field(None, description="Acceptable confidence levels")


# Batch Processing Schemas
class BatchMatchRequest(BaseModel):
    order_ids: List[int] = Field(..., min_items=1, max_items=50, description="List of order IDs to match")
    limit_per_order: int = Field(5, ge=1, le=20, description="Maximum matches per order")
    min_score: float = Field(0.6, ge=0.0, le=1.0, description="Minimum match score threshold")
    filters: Optional[MatchFilters] = Field(None, description="Additional filtering criteria")


# Analytics Schemas
class CategoryMatchStats(BaseModel):
    category: str
    match_count: int
    average_score: Optional[float] = None
    conversion_rate: Optional[float] = None


class MatchAnalytics(BaseModel):
    total_matches_generated: int = Field(..., description="Total matches generated")
    successful_connections: int = Field(..., description="Matches that led to connections")
    average_match_score: float = Field(..., ge=0.0, le=1.0, description="Average match score")
    top_matching_categories: List[Dict[str, Any]] = Field(..., description="Top performing categories")
    conversion_rate: float = Field(..., ge=0.0, le=1.0, description="Overall conversion rate")
    average_response_time_hours: float = Field(..., gt=0, description="Average response time in hours")
    user_satisfaction_score: float = Field(..., ge=0.0, le=5.0, description="User satisfaction rating")
    match_quality_trends: Optional[List[Dict[str, Any]]] = Field(None, description="Match quality over time")
    geographic_distribution: Optional[Dict[str, int]] = Field(None, description="Matches by geographic region")


# Feedback Schema
class MatchFeedback(BaseModel):
    match_id: str = Field(..., description="ID of the match being rated")
    feedback_type: FeedbackType = Field(..., description="Type of feedback")
    rating: Optional[int] = Field(None, ge=1, le=5, description="Rating from 1-5")
    comment: Optional[str] = Field(None, max_length=500, description="Optional feedback comment")
    contacted_manufacturer: bool = Field(False, description="Whether user contacted the manufacturer")
    resulted_in_quote: bool = Field(False, description="Whether it resulted in a quote")
    resulted_in_order: bool = Field(False, description="Whether it resulted in an order")


# Real-time Matching Schemas
class LiveMatchRequest(BaseModel):
    order_id: int = Field(..., description="Order ID for live matching")
    include_capacity_check: bool = Field(True, description="Include real-time capacity checking")
    priority_boost: Optional[float] = Field(None, ge=1.0, le=3.0, description="Priority boost multiplier")


class LiveMatchResponse(SmartMatchResponse):
    real_time_capacity: Optional[Dict[str, Any]] = Field(None, description="Real-time capacity information")
    availability_window: Optional[Dict[str, str]] = Field(None, description="Current availability window")
    response_time_estimate: Optional[int] = Field(None, description="Estimated response time in minutes")


# Recommendation Schemas
class RecommendationRequest(BaseModel):
    user_preferences: Optional[Dict[str, Any]] = Field(None, description="User preference overrides")
    include_historical: bool = Field(True, description="Include historical performance data")
    boost_new_manufacturers: bool = Field(False, description="Boost newer manufacturers for diversity")
    limit: int = Field(10, ge=1, le=50, description="Maximum number of recommendations")


class PersonalizedRecommendation(SmartMatchResponse):
    recommendation_reason: str = Field(..., description="Why this was recommended")
    historical_performance: Optional[Dict[str, Any]] = Field(None, description="Historical performance data")
    similarity_score: Optional[float] = Field(None, ge=0.0, le=1.0, description="Similarity to past successful matches")


# Advanced Matching Schemas
class MLMatchingInsights(BaseModel):
    predicted_success_rate: float = Field(..., ge=0.0, le=1.0, description="ML predicted success rate")
    key_success_factors: List[str] = Field(..., description="Key factors for success")
    risk_factors: List[str] = Field(..., description="Potential risk factors")
    similar_successful_matches: Optional[List[str]] = Field(None, description="Similar successful match IDs")
    confidence_interval: Optional[Dict[str, float]] = Field(None, description="Confidence interval for predictions")


class EnhancedMatchResponse(SmartMatchResponse):
    ml_insights: Optional[MLMatchingInsights] = Field(None, description="Machine learning insights")
    market_context: Optional[Dict[str, Any]] = Field(None, description="Current market context")
    competitive_analysis: Optional[Dict[str, Any]] = Field(None, description="Competitive positioning")


# Notification Schemas
class MatchNotification(BaseModel):
    notification_id: str = Field(..., description="Unique notification ID")
    match_id: str = Field(..., description="Associated match ID")
    recipient_id: int = Field(..., description="User ID to notify")
    notification_type: str = Field(..., description="Type of notification")
    title: str = Field(..., description="Notification title")
    message: str = Field(..., description="Notification message")
    priority: str = Field("normal", description="Notification priority")
    created_at: datetime = Field(..., description="When notification was created")
    expires_at: Optional[datetime] = Field(None, description="When notification expires")
    action_url: Optional[str] = Field(None, description="URL for notification action")


# Export Schemas
class MatchExportRequest(BaseModel):
    match_ids: Optional[List[str]] = Field(None, description="Specific match IDs to export")
    date_range: Optional[Dict[str, str]] = Field(None, description="Date range for export")
    format: str = Field("csv", pattern="^(csv|excel|pdf)$", description="Export format")
    include_analytics: bool = Field(True, description="Include analytics data")
    include_feedback: bool = Field(False, description="Include feedback data")


class MatchExportResponse(BaseModel):
    export_id: str = Field(..., description="Export job ID")
    status: str = Field(..., description="Export status")
    download_url: Optional[str] = Field(None, description="Download URL when ready")
    created_at: datetime = Field(..., description="Export creation time")
    estimated_completion: Optional[datetime] = Field(None, description="Estimated completion time")
    record_count: Optional[int] = Field(None, description="Number of records to export")

File: frontend/src/lib/api/smartMatching.ts
import api from '../api';
import { 
  SmartMatch, 
  MatchFilters, 
  MatchAnalytics, 
  MatchFeedback,
  BatchMatchRequest,
  LiveMatchRequest,
  RecommendationRequest
} from '../../types';

export interface SmartMatchResponse {
  match_id: string;
  match_type: 'order_to_production_quote' | 'production_quote_to_order';
  order_id?: number;
  production_quote_id?: number;
  score: {
    total_score: number;
    category_match: number;
    price_compatibility: number;
    timeline_compatibility: number;
    geographic_proximity: number;
    capacity_availability: number;
    manufacturer_rating: number;
    urgency_alignment: number;
    specification_match: number;
    confidence_level: 'EXCELLENT' | 'VERY_GOOD' | 'GOOD' | 'FAIR' | 'POOR';
    match_reasons: string[];
    potential_issues: string[];
  };
  estimated_price?: number;
  estimated_delivery_days?: number;
  manufacturer_info: {
    id: number;
    name: string;
    location?: string;
    rating?: number;
    verified: boolean;
    completed_orders: number;
  };
  created_at: string;
  expires_at?: string;
  real_time_availability?: string;
}

export interface MatchAnalyticsResponse {
  total_matches_generated: number;
  successful_connections: number;
  average_match_score: number;
  top_matching_categories: Array<{
    category: string;
    match_count: number;
    average_score?: number;
    conversion_rate?: number;
  }>;
  conversion_rate: number;
  average_response_time_hours: number;
  user_satisfaction_score: number;
  match_quality_trends?: Array<{
    date: string;
    average_score: number;
    match_count: number;
  }>;
  geographic_distribution?: Record<string, number>;
}

export const smartMatchingApi = {
  // Core Matching Functions
  async getMatchesForOrder(
    orderId: number, 
    options: {
      limit?: number;
      minScore?: number;
    } = {}
  ): Promise<SmartMatchResponse[]> {
    const params = new URLSearchParams();
    if (options.limit) params.append('limit', options.limit.toString());
    if (options.minScore) params.append('min_score', options.minScore.toString());
    
    return api.get(`/smart-matching/orders/${orderId}/matches?${params}`);
  },

  async getMatchesForProductionQuote(
    quoteId: number,
    options: {
      limit?: number;
      minScore?: number;
    } = {}
  ): Promise<SmartMatchResponse[]> {
    const params = new URLSearchParams();
    if (options.limit) params.append('limit', options.limit.toString());
    if (options.minScore) params.append('min_score', options.minScore.toString());
    
    return api.get(`/smart-matching/production-quotes/${quoteId}/matches?${params}`);
  },

  // Batch Processing
  async batchMatchOrders(request: {
    order_ids: number[];
    limit_per_order?: number;
    min_score?: number;
    filters?: MatchFilters;
  }): Promise<SmartMatchResponse[]> {
    return api.post('/smart-matching/batch-match', request);
  },

  // Personalized Recommendations
  async getPersonalizedRecommendations(options: {
    limit?: number;
    match_type?: 'order_to_production_quote' | 'production_quote_to_order';
  } = {}): Promise<SmartMatchResponse[]> {
    const params = new URLSearchParams();
    if (options.limit) params.append('limit', options.limit.toString());
    if (options.match_type) params.append('match_type', options.match_type);
    
    return api.get(`/smart-matching/recommendations?${params}`);
  },

  // Real-time Matching
  async getLiveMatches(orderId: number): Promise<SmartMatchResponse[]> {
    return api.get(`/smart-matching/live-matches/${orderId}`);
  },

  async getPriorityMatches(
    orderId: number, 
    urgencyBoost: number = 1.5
  ): Promise<SmartMatchResponse[]> {
    return api.post(`/smart-matching/priority-matching?order_id=${orderId}&urgency_boost=${urgencyBoost}`);
  },

  // Analytics
  async getMatchingAnalytics(days: number = 30): Promise<MatchAnalyticsResponse> {
    // Ensure auth credentials (cookies or token header) are included – this endpoint requires manufacturer/admin scope
    const token = localStorage.getItem('accessToken') || localStorage.getItem('auth_token');
    const config = token
      ? { withCredentials: true as const, headers: { Authorization: `Bearer ${token}` } }
      : { withCredentials: true as const };
    return api.get(`/smart-matching/analytics?days=${days}`, config);
  },

  // Feedback
  async submitMatchFeedback(feedback: {
    match_id: string;
    feedback_type: 'helpful' | 'not_helpful' | 'contacted' | 'converted' | 'irrelevant';
    rating?: number;
    comment?: string;
    contacted_manufacturer?: boolean;
    resulted_in_quote?: boolean;
    resulted_in_order?: boolean;
  }): Promise<{ message: string; match_id: string; feedback_type: string }> {
    return api.post('/smart-matching/feedback', feedback);
  },

  // Cache Management
  async refreshCache(): Promise<{ message: string }> {
    return api.post('/smart-matching/refresh-cache');
  },

  // Health Check
  async getHealthStatus(): Promise<{
    status: string;
    cache_size: number;
    service: string;
    version: string;
  }> {
    return api.get('/smart-matching/health');
  }
};

// Helper Functions for Smart Matching
export const smartMatchingHelpers = {
  // Score Interpretation
  getScoreColor(score: number): string {
    if (score >= 0.9) return 'green';
    if (score >= 0.8) return 'blue';
    if (score >= 0.7) return 'yellow';
    if (score >= 0.6) return 'orange';
    return 'red';
  },

  getScoreLabel(score: number): string {
    if (score >= 0.9) return 'Excellent Match';
    if (score >= 0.8) return 'Very Good Match';
    if (score >= 0.7) return 'Good Match';
    if (score >= 0.6) return 'Fair Match';
    return 'Poor Match';
  },

  getConfidenceColor(confidence: string): string {
    switch (confidence) {
      case 'EXCELLENT': return 'green';
      case 'VERY_GOOD': return 'blue';
      case 'GOOD': return 'yellow';
      case 'FAIR': return 'orange';
      case 'POOR': return 'red';
      default: return 'gray';
    }
  },

  getConfidenceLabel(confidence: string): string {
    switch (confidence) {
      case 'EXCELLENT': return 'Excellent';
      case 'VERY_GOOD': return 'Very Good';
      case 'GOOD': return 'Good';
      case 'FAIR': return 'Fair';
      case 'POOR': return 'Poor';
      default: return 'Unknown';
    }
  },

  // Match Analysis
  analyzeMatchStrengths(match: SmartMatchResponse): string[] {
    const strengths = [];
    const score = match.score;

    if (score.category_match >= 0.9) {
      strengths.push('Perfect category alignment');
    }
    if (score.price_compatibility >= 0.8) {
      strengths.push('Competitive pricing');
    }
    if (score.timeline_compatibility >= 0.8) {
      strengths.push('Meets delivery timeline');
    }
    if (score.manufacturer_rating >= 0.8) {
      strengths.push('Highly rated manufacturer');
    }
    if (score.capacity_availability >= 0.8) {
      strengths.push('Good availability');
    }
    if (score.geographic_proximity >= 0.8) {
      strengths.push('Favorable location');
    }

    return strengths;
  },

  analyzeMatchWeaknesses(match: SmartMatchResponse): string[] {
    const weaknesses = [];
    const score = match.score;

    if (score.price_compatibility < 0.5) {
      weaknesses.push('Pricing concerns');
    }
    if (score.timeline_compatibility < 0.5) {
      weaknesses.push('Timeline challenges');
    }
    if (score.manufacturer_rating < 0.5) {
      weaknesses.push('Limited manufacturer track record');
    }
    if (score.capacity_availability < 0.5) {
      weaknesses.push('Limited capacity');
    }
    if (score.geographic_proximity < 0.5) {
      weaknesses.push('Distant location');
    }

    return weaknesses;
  },

  // Recommendation Logic
  shouldRecommendMatch(match: SmartMatchResponse): boolean {
    return match.score.total_score >= 0.7 && 
           match.score.confidence_level !== 'POOR' &&
           match.manufacturer_info.verified;
  },

  getPriorityLevel(match: SmartMatchResponse): 'high' | 'medium' | 'low' {
    if (match.score.total_score >= 0.85) return 'high';
    if (match.score.total_score >= 0.7) return 'medium';
    return 'low';
  },

  // Formatting Helpers
  formatMatchScore(score: number): string {
    return `${Math.round(score * 100)}%`;
  },

  formatEstimatedPrice(price?: number, currency: string = 'USD'): string {
    if (!price) return 'Contact for pricing';
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: currency
    }).format(price);
  },

  formatDeliveryTime(days?: number): string {
    if (!days) return 'TBD';
    if (days === 1) return '1 day';
    if (days < 7) return `${days} days`;
    if (days < 30) return `${Math.round(days / 7)} weeks`;
    return `${Math.round(days / 30)} months`;
  },

  formatManufacturerRating(rating?: number): string {
    if (!rating) return 'Not rated';
    return `${rating.toFixed(1)}/5.0`;
  },

  // Sorting and Filtering
  sortMatchesByScore(matches: SmartMatchResponse[]): SmartMatchResponse[] {
    return [...matches].sort((a, b) => b.score.total_score - a.score.total_score);
  },

  sortMatchesByPrice(matches: SmartMatchResponse[]): SmartMatchResponse[] {
    return [...matches].sort((a, b) => {
      const priceA = a.estimated_price || Infinity;
      const priceB = b.estimated_price || Infinity;
      return priceA - priceB;
    });
  },

  sortMatchesByDelivery(matches: SmartMatchResponse[]): SmartMatchResponse[] {
    return [...matches].sort((a, b) => {
      const deliveryA = a.estimated_delivery_days || Infinity;
      const deliveryB = b.estimated_delivery_days || Infinity;
      return deliveryA - deliveryB;
    });
  },

  filterMatchesByScore(matches: SmartMatchResponse[], minScore: number): SmartMatchResponse[] {
    return matches.filter(match => match.score.total_score >= minScore);
  },

  filterMatchesByVerified(matches: SmartMatchResponse[]): SmartMatchResponse[] {
    return matches.filter(match => match.manufacturer_info.verified);
  },

  filterMatchesByPrice(matches: SmartMatchResponse[], maxPrice: number): SmartMatchResponse[] {
    return matches.filter(match => 
      !match.estimated_price || match.estimated_price <= maxPrice
    );
  },

  // Match Comparison
  compareMatches(matchA: SmartMatchResponse, matchB: SmartMatchResponse): {
    better_match: SmartMatchResponse;
    comparison_points: string[];
  } {
    const points = [];
    let betterMatch = matchA;

    if (matchA.score.total_score > matchB.score.total_score) {
      points.push(`Match A has higher overall score (${this.formatMatchScore(matchA.score.total_score)} vs ${this.formatMatchScore(matchB.score.total_score)})`);
    } else if (matchB.score.total_score > matchA.score.total_score) {
      points.push(`Match B has higher overall score (${this.formatMatchScore(matchB.score.total_score)} vs ${this.formatMatchScore(matchA.score.total_score)})`);
      betterMatch = matchB;
    }

    if (matchA.estimated_price && matchB.estimated_price) {
      if (matchA.estimated_price < matchB.estimated_price) {
        points.push(`Match A is more affordable (${this.formatEstimatedPrice(matchA.estimated_price)} vs ${this.formatEstimatedPrice(matchB.estimated_price)})`);
      } else {
        points.push(`Match B is more affordable (${this.formatEstimatedPrice(matchB.estimated_price)} vs ${this.formatEstimatedPrice(matchA.estimated_price)})`);
      }
    }

    if (matchA.estimated_delivery_days && matchB.estimated_delivery_days) {
      if (matchA.estimated_delivery_days < matchB.estimated_delivery_days) {
        points.push(`Match A delivers faster (${this.formatDeliveryTime(matchA.estimated_delivery_days)} vs ${this.formatDeliveryTime(matchB.estimated_delivery_days)})`);
      } else {
        points.push(`Match B delivers faster (${this.formatDeliveryTime(matchB.estimated_delivery_days)} vs ${this.formatDeliveryTime(matchA.estimated_delivery_days)})`);
      }
    }

    return {
      better_match: betterMatch,
      comparison_points: points
    };
  },

  // Analytics Helpers
  calculateMatchSuccessRate(analytics: MatchAnalyticsResponse): number {
    if (analytics.total_matches_generated === 0) return 0;
    return analytics.successful_connections / analytics.total_matches_generated;
  },

  getTopPerformingCategory(analytics: MatchAnalyticsResponse): string | null {
    if (analytics.top_matching_categories.length === 0) return null;
    return analytics.top_matching_categories[0].category;
  },

  // Real-time Updates
  isMatchExpired(match: SmartMatchResponse): boolean {
    if (!match.expires_at) return false;
    return new Date(match.expires_at) < new Date();
  },

  getMatchAge(match: SmartMatchResponse): string {
    const created = new Date(match.created_at);
    const now = new Date();
    const diffMs = now.getTime() - created.getTime();
    const diffMins = Math.floor(diffMs / 60000);
    const diffHours = Math.floor(diffMins / 60);
    const diffDays = Math.floor(diffHours / 24);

    if (diffMins < 1) return 'Just now';
    if (diffMins < 60) return `${diffMins}m ago`;
    if (diffHours < 24) return `${diffHours}h ago`;
    return `${diffDays}d ago`;
  },

  // Export Helpers
  exportMatchesToCSV(matches: SmartMatchResponse[]): string {
    const headers = [
      'Match ID',
      'Type',
      'Score',
      'Confidence',
      'Manufacturer',
      'Estimated Price',
      'Delivery Days',
      'Created At'
    ];

    const rows = matches.map(match => [
      match.match_id,
      match.match_type,
      this.formatMatchScore(match.score.total_score),
      match.score.confidence_level,
      match.manufacturer_info.name,
      this.formatEstimatedPrice(match.estimated_price),
      this.formatDeliveryTime(match.estimated_delivery_days),
      new Date(match.created_at).toLocaleDateString()
    ]);

    return [headers, ...rows]
      .map(row => row.map(cell => `"${cell}"`).join(','))
      .join('\n');
  }
}; 